File: oop.js

// Choose one version
var classes = require("./oop_v1.js");
//var classes = require("./oop_v2.js");
//var classes = require("./oop_v3.js");


// Extract objects from export module
var Machine = classes.Machine;
var Tv = classes.Tv;


console.log("\n-- create instances --");

var m = new Machine("simple machine", 10);
m.describe();

var tv = new Tv(1000, "LED");
tv.setVolume(50);
tv.describe();

console.log("\n-- check types --");

// You can check the type of an instance
console.log("tv is a Tv ? ", tv instanceof Tv);
console.log("tv is a Machine ? ", tv instanceof Machine);
console.log("m is a Tv ? ", m instanceof Tv);
console.log("m is a Machine ? ", m instanceof Machine);

console.log("");

File: oop.rb

class Machine

	def initialize(name, price)
		@name = name
		@price = price
	end

	def describe
		puts "This is a #{self}"
	end

	def to_s
		"#{@name}, costs #{@price} €"
	end
end

class Tv < Machine

	attr_accessor :volume 

	def initialize(price, type)
		super("television", price)
		@type = type
		@volume = 0
	end

	def to_s
		"#{super.to_s}, type #{@type}, current volume #{@volume}"
	end
end


# Use of classes

puts "\n-- create instances --"

m = Machine.new("simple machine", 10)
m.describe

tv = Tv.new(1000, "LED")
tv.volume = 50
tv.describe

puts "\n-- check types --"

puts "tv is a Tv ? #{tv.is_a? Tv}"
puts "tv is a Machine ? #{tv.is_a? Machine}" 
puts "m is a Tv ? #{m.is_a? Tv}"
puts "m is a Machine ? #{m.is_a? Machine}"
 
puts ""
